File redirection

lierdakil opened the pull request:
## THIS IS WIP, DON'T MERGE

As discussed in #482, I started working on this. Initial results show promise.

I want to run this through Travis to see if I broke anything.

This is a working prototype, but code style is rather horrible (sleep deprivation does that)

I would be very grateful if someone reviews my efforts thus far.

TODO:
- [x] Cleanup
- [x] Load all mapped files as targets
- [x] Replace mapped names with original ones in output
- [ ] Verification
- [x] Tests
- [x] Bugfixes
- [ ] Docs
- [ ] Squash?

lierdakil commented:
Ha. Apparently, I've made GHC crash. Oh joy.

lierdakil commented:
I think I managed to work around GHC 7.4 bugs, and cleaned up code somewhat in the process. Would be great if anyone who knows their way around GHC-Mod could check if I encroach on something I shouldn't. I think that should be fine, but I'm not an expert.

DanielG commented on src/GHCMod.hs:
```diff
@@ -257,6 +258,14 @@ globalArgSpec =
                reqArg "OPT" $ \g o ->
                    o { ghcUserOptions = g : ghcUserOptions o }
 
+      , option "m" ["file-map"] "Redirect one file to another, -m \"file1.hs=file2.hs\"" $
```
Since this is mostly backend option I'd rather only allocate a logopt for it.

DanielG commented on Language/Haskell/GhcMod/Monad/Types.hs:
```diff
@@ -129,10 +138,11 @@ data GhcModState = GhcModState {
       gmGhcSession   :: !(Maybe GmGhcSession)
     , gmComponents   :: !(Map ChComponentName (GmComponent GMCResolved (Set ModulePath)))
     , gmCompilerMode :: !CompilerMode
+    , gmMMappedFiles :: Map TargetId (StringBuffer, UTCTime)
```
I'm not convinced using TargetId as the key is a good idea. Since a `TargetModule` might point to the same file as a `TargetFile` so it's not unique which would defeat the point of the map. You should probably just store the module name and use HomeModuleGraph to look up the corresponding file at the last minute if that's needed.

lierdakil replied:
I'm not sure how that would work. As I see it, there are three options:

1. Map keys can only be filenames
2. Map keys can only be module names
3. Map keys can be either.

For (1) and (2) this is a moot point, and TargetId is just a convenience choice here, since only one constructor is ever used. For (3), it doesn't really matter if it's TargetId or String or otherwise, since duplication is possible in any case, and I don't see how I would extract module name given source code and file name, assuming there can be errors in said code.

DanielG replied:
Maybe I didn't explain this properly. The problem is that guessTarget can return either a module name or a file name but you use Eq/Ord on the TargetId type when finding out if a target is mapped. So if guessTarget returns a module name like say `Language.GhcMod.Target` but then the user/editor frontend passes ghc-mod the filename of that module the mapping wont apply.

So you need to find a way to either map module names onto absolute paths or paths onto module names. Usually the former is done using the ModuleGraph from ghc's `depanal` but since depanal won't honor the file remapping that might not be an option.

So instead you can modify [HomeModuleGraph](https://github.com/kazu-yamamoto/ghc-mod/blob/master/Language/Haskell/GhcMod/HomeModuleGraph.hs) to also honor your mapping. Right now HMG is not meant to be absolutely correct since it's "only" used to find out which modules/files belong to which cabal components. So errors coming from it are mostly ignored since we'll still get a partial module graph that might work. The infrastructure for logging and getting out errors and warnings is there though just needs to be actually used.

HMG is fairly forgiving already, it only parses the "module head" i.e. `module X where` and import statements any parse errors after which will not disturb it. Only processors are a major pain and will need special handling.

lierdakil replied:
Ok, I think I kind of understand the idea now. A couple points you might have missed though (or where I lack understanding maybe):

1. Imports don't generally honor mappings, and only way to make this happen is load all mapped files as targets, which doesn't sound that great. I don't think this is a *huge* problem, since, in general, it only makes sense to map a couple files that are actually being passed as targets, and not whole project.
2. Only command where ghc-mod accepts module name is `browse` I think, so for other commands `TargetFile` is looked up as `TargetFile`
3. `guessTarget` favors `TargetFile` greatly: if it can interpret a given `String` as a filename, it will.

So from where I stand, it seems like concerns would be that `browse` and `HomeModuleGraph` don't honor mappings.

One way to fix this would be to parse module name from source files when loading mappings and store those in `Map ModuleName ...`, and then get filenames from HMG when loading targets. But that doesn't strike me as the most straightforward option, since, in general, we already know filenames when loading mapped files.

Maybe a better option would be to add *both* `TargetFile` and `TargetModule` to `gmMMappedFIles` when loading those. This way, we save on an HMG trip at a cost of a little more memory, and that would work in most cases even if it's impossible to determine module name due to errors.

Does that sound reasonable to you?

lierdakil replied:
Scratch that about having both in `gmMMappedFIles`, that's not very feasible. But I still maintain that matching on canonical filenames is more robust than matching on module names, since latter is not always well-defined.

DanielG replied:
> Ok, I think I kind of understand the idea now. A couple points you
> might have missed though (or where I lack understanding maybe):

> 1. Imports don't generally honor mappings, 

This I don't understand, how would any of this work if an import
didn't honor the mapping??

> and only way to make this
> happen is load all mapped files as targets, which doesn't sound that
> great.

But that's exactly what you have to do to get ghc to accept your
remapped module, is it not? .. I'm so confused.

> I don't think this is a *huge* problem, since, in general, it
> only makes sense to map a couple files that are actually being passed
> as targets, and not whole project.

Well that depends entirely on whay you are doing. If you want to build
a cloud IDE or something based on ghc-mod you'd want to be able to map
the whole project so you don't need to write modules out to the
filesystem maybe.

> 2. Only command where ghc-mod accepts module name is `browse` I
> think, so for other commands `TargetFile` is looked up as `TargetFile`

That's not quite true module names are accepted all over the place.
Most notably also by `check`. Anything that uses `GmlT` can use either
module or file names.

> 3. `guessTarget` favors `TargetFile` greatly: if it can interpret a
> given `String` as a filename, it will.

But only if the file actually exists IIRC from the docs.

> So from where I stand, it seems like concerns would be that `browse`
> and `HomeModuleGraph` don't honor mappings.

> One way to fix this would be to parse module name from source files
> when loading mappings and store those in `Map ModuleName ...`,

I already do that anyways, you just need to add support for remapping.
HMG deals exclusively in fully resolved Module+FilePath units
(`ModulePath`). `resolveModule` implements the FilePath -> ModulePath
mapping and it already does use `fileModuleName` to parse the module
and extract the module name. So your remapping code would have to go
somewhere between `resolveModule` and `updateHomeModuleGraph`.

> and
> then get filenames from HMG when loading targets. But that doesn't
> strike me as the most straightforward option, since,

> in general, we already know filenames when loading mapped files.

We can't just ignore those corner cases though. ghc-mod has to work in
any and all circumstances (within reason).

> [...]

> Scratch that about having both in gmMMappedFiles, that's not very
> feasible. But I still maintain that matching on canonical filenames is
> more robust than matching on module names, since latter is not always
> well-defined.

Let's enumerate ~~all~~ some of the possible cases:

Assuming we have a project with the following module graph:

    A <- B <- D
     \       /
      C <----

With the filesystem layout being the following

    proj.cabal
    `- src
       `- A.hs
        - B.hs
        - C.hs
        - D.hs

Let's look at a few different mappings given the following scenarios:

- The user references module `A` but src/A.hs has a preprocessor error.

    1) module A -> /tmp/A.hs

       That's fine since the user referenced that exact module name
       which we can match on.

    2) src/A.hs -> /tmp/A.hs

       That's fine since when looking for a module called `A` we can use ghc's
       search path algorithm. We will eventually look in src/ since it ought
       to be on the search path. And we'll find a file there matching the
       module name convention therefore that must be the module A.

- The user references the file `src/A.hs` directly but again the
  original has a preprocessor error

   1) module A -> /tmp/A.hs

      This case is more tricky. We have maybe a few possibilities here:

      - We could try to open src/A.hs and parse it's module name but
        as you not this is not always well defined.

      - Apply the search path algorithm in reverse and realize the fact
        that since src/ is on the search path the file called A.hs in this
        directory must be the module A! Otherwise ghc wouldn't find it when
        it's looking for it anyways. Now this would be a bit finicky since
        you have to be careful with the order of directories on the search
        path but I imagine this is feasible.

   2) src/A.hs -> /tmp/A.hs

      This is fine since we can just match on the canonicalized path.

Can you think of any others?

lierdakil replied:
I will answer the main bulk of your comment probably tomorrow, since I'm pretty sleepy now, but a couple more facts that you may not be aware of:
 - GHC API requires a file to exist when loading targets. Its contents can be supplied via argument, but file itself *has* to be there, as far as I can tell, although it can be empty, or have arbitrary contents.
 - GHC API forbids CPP preprocessing on in-memory files (more like, it just can't handle it on its own, actually). It's generally possible to query for types etc, but when running a check, it just bails.

So no, building a could IDE based just on this mechanic is not feasible. It's handy when running a type query on a modified-but-not-yet-saved file, but that's pretty much the limit of its usefulness.

> But that's exactly what you have to do to get ghc to accept your remapped module, is it not? .. I'm so confused.

For now, I just load mappings that are explicitly passed as targets, not *all* mapped files (i.e. check file.hs would only load a mapping for file.hs, if it's specified -- for my use-case that's actually quite enough). But you have a point there -- if there is no need to load a file, there is no reason to define a mapping for it, so I should probably reevaluate that.

> But only if the file actually exists IIRC from the docs.
As I said, GHC API requires file to actually exist to supply a source for it, but even if corresponding file doesn't exist, if a target "looks" like filename, i.e. has `.hs`/`.lhs` suffix, then `guessTarget` interprets it as filename regardless. Also if adding `.hs` or `.lhs` to argument string yields an existing filename, it will be interpreted as filename. If everything else fails, only then it would be interpreted as a module name (that is, if it doesn't contain characters forbidden in module name, otherwise it just throws)

lierdakil replied:
Sorry about the hold-up, I caught cold somehow.

On the topic of possible options, after thinking about this for a while, I think `--file-map` should only accept filenames on both left and right, at least for now. This would eliminate trickier corner cases, while being reasonably useful. With this limitation, it becomes somewhat easier.

Guessing if it's filename or module name is relatively cumbersome and error-prone, it could be a better idea in general to add another option for modules altogether.

FYI, on `master` branch of ghc-mod, `ghc-mod check ModuleName` doesn't work: it throws `canonicalizePath: does not exist (No such file or directory)`.

DanielG replied:
Sure just implement what you need for now. We can always extend that later.

lierdakil commented:
@DanielG, thank you for your comments, I will be sure to follow your advice.

I found a more serious problem with my current approach though, w.r.t. supplying files on stdin (`MemoryMapping`). Basically, I read from stdin after end, which is all fun and good as long as I end stream with Ctrl+D from console. But most languages frown on *writing* into a stream after end, so it's not very nice when only way to pass file in interactive mode is close the stream.

Only option I see (short of using temporary files/named pipes/sockets) is using surrogate EOF character, but I'm not entirely sure what to use. Semantically, that might be either one of `\ETX` (end of text), `\EOT` (end of transmission), `\ETB` (end of transmission block), `\FS` (file separator). Since I don't use Emacs, I have no idea if there is a preference or maybe entirely different option that would be more convenient. Could you provide some insight on this? If possible, I'd prefer to avoid using sockets/named pipes, since I feel like it introduces unnecessary complexity.

lierdakil commented:
Okay, this could use another review, if possible.

DanielG commented:
Argh, sorry I completely forgot about this :/

DanielG commented on Language/Haskell/GhcMod/FileMapping.hs:
```diff
@@ -0,0 +1,67 @@
+  addToState from fm
+loadMappedFile from (MemoryMapping _) = do
+  let loop' acc = do
+        line <- getLine
```
This seems the wrong place to do that piece of IO. Should really be in the frontend code.

DanielG commented on Language/Haskell/GhcMod/HomeModuleGraph.hs:
```diff
@@ -206,17 +208,33 @@ updateHomeModuleGraph' env smp0 = do
+          (dflags, src) <- case mapping of
+            Just (RedirectedMapping to) -> preprocess'' to
+            Just (MemoryMapping (Just src)) -> do
+              (fp, hndl) <- liftIO $ getTemporaryDirectory >>= \tmpDir' ->
```
Please re-use the tempdir in the `Cradle` type. You'll have to pass it down to this function but I don't want ghc-mod to create a hundred tempdirs per run.

lierdakil replied:
Actually, getTemporaryDirectory just returns path to `/tmp` or `C:\Temp` or something. But I get your point, better use cradleTempDir anyway.

DanielG commented on Language/Haskell/GhcMod/Logger.hs:
```diff
@@ -33,6 +36,13 @@ data Log = Log [String] Builder
 
 newtype LogRef = LogRef (IORef Log)
 
+data ReaderState = ReaderState { rsDynFlags :: DynFlags
```
A less generic name would be nice, maybe GmPprEnv or something.

DanielG commented on Language/Haskell/GhcMod/Logger.hs:
```diff
@@ -33,6 +36,13 @@ data Log = Log [String] Builder
+                               , rsMapFile :: FilePath -> FilePath
+                               }
+
+type ReaderM a = Reader ReaderState a
```
Same here

DanielG commented on Language/Haskell/GhcMod/Logger.hs:
```diff
@@ -70,67 +80,90 @@ withLogger f action = do
+        }
+
+        setLogger df = Gap.setLogAction df $ appendLogRef st df logref
+        handlers = [
```
Can you elaborate on what these handlers do?

DanielG commented on Language/Haskell/GhcMod/Target.hs:
```diff
@@ -151,11 +154,23 @@ runGmlTWith efnmns' mdf wrapper action = do
+  where
+    relativize (Target (TargetFile filePath phase) taoc src) = do
+      crdl <- cradle
+      let tid = makeRelative (cradleRootDir crdl) filePath `TargetFile` phase
```
This seems like a really weird place for an infix call, don't do that.

lierdakil replied:
I just don't like endless parenthesis. Oh well.

DanielG commented on src/GHCMod.hs:
```diff
@@ -399,6 +408,12 @@ legacyInteractiveLoop symdbreq ref world = do
         "boot"   -> bootCmd []
         "browse" -> browseCmd args
 
+        "load"   -> loadMappedFile arg (MemoryMapping Nothing)
```
`map-file` and `unmap-file` seem more appropriate names for the commands.

DanielG commented on src/GHCMod.hs:
```diff
@@ -256,6 +257,14 @@ globalArgSpec =
                reqArg "OPT" $ \g o ->
                    o { ghcUserOptions = g : ghcUserOptions o }
 
+      , option "" ["file-map"] "Redirect one file to another, --file-map \"file1.hs=file2.hs\"" $
```
`file-map` seems backwards, `map-file` sounds more like a command.

DanielG commented:
Looks pretty good so far :)

DanielG commented:
Awesome, just add some tests and maybe documentation and we're good to go.

lierdakil commented:
I wasn't in any shape to work on this for the past couple of weeks, so all is good.

I'll try to write some tests by the end of this week.

lierdakil commented:
Ok, I found out that linting doesn't work on mapped files. To avoid using temporary files, might be a good idea to switch to newer hlint (hlint3) API, that allows to pass file source as an argument to `hlint`.

Also, looking at hlint sources, it might be a good idea to use [CppHs](https://hackage.haskell.org/package/cpphs-1.19/docs/Language-Preprocessor-Cpphs.html) for preprocessing without need for temp. files (hlint depends on it anyway, so it's not strictly an additional dependency)

What do you think?

DanielG commented:
Sounds good.

lierdakil commented:
After playing with it for a bit, I think I'll hold off on that, and just use temp. files for now, to keep scope of this PR manageable. I'll create other PRs for newer hlint API and for CppHs, when I get to it.

DanielG commented:
That's fine too.

lierdakil commented:
Okay, added some tests, but two of those fail (that's why one doesn't rely on manual testing...)

One because `info` shows temp. file name, that I know how to fix, just didn't notice that until now. 

Another seems more complicated. Apparently, `type` crashes on redirected files (on in-memory files it's perfectly fine). I'll try to track it down during weekend.

DanielG commented:
Awesome :)

lierdakil commented:
I managed to fix both failing tests. I'm a little bit unsure about changing interfaces to `Gap.pprInfo` and `Gap.infoThing`. If those are exported (not sure if they are), that would be API-breaking change. I could add stubs conforming to old interface if that's needed.

lierdakil commented:
Ping. I could use some pointers on where to add docs on this exactly. Also please see if there's anything you'd like to change about current code.

lierdakil commented:
Hello? @DanielG

DanielG commented:
I'm really sorry I just don't have any time right now have to finish stuff for work before I go to Japan. In about a week and a bit I'll be working on ghc-mod full time so expect a review then :)

lierdakil commented:
Ok great.

DanielG commented on Language/Haskell/GhcMod/Utils.hs:
```diff
@@ -156,3 +160,41 @@ canonFilePath f = do
+      return result
+    runWithFile _ = action file
+
+getCanonicalFileNameSafe :: (IOish m, GmEnv m) => FilePath -> m FilePath
```
This seems like a bad idea, once the file gets created (possibly by the user) the "canonical" name could change you'll have to comment on all call sites to convince me this is safe.

lierdakil replied:
A fair point. I can probably do away with just `ccfn`, dropping `canonicalizePath`. Would that address your concern?

lierdakil replied:
On second thought, this will be a problem with HMG, since it stores canonical filenames.

DanielG replied:
So as far as I understand you need this because the redirected file might not actually exist yet, right? I don't know if that would be a good idea but how about just requiring it to already exist? (Like even just as an empty file)

lierdakil replied:
It's an arbitrary limitation with no convincing reason, so not a fan of this particular idea.

In all honesty, for my intended use-case, possibility of changing filesystem shouldn't be a problem, since I actually 'unmap' files once I'm done with them, so unless filesystem changes right in the middle of a command (which is not very likely), we're good. But I can construct a test case where this might actually fail (that would need symlinks though -- so not on Windows).

I think I might be able to defer canonicalization until file is actually used, which would somewhat mitigate this issue.

DanielG commented on Language/Haskell/GhcMod/Utils.hs:
```diff
@@ -156,3 +160,41 @@ canonFilePath f = do
+    then liftIO $ canonicalizePath ccfn
+    else return ccfn
+
+mkRevRedirMapFunc :: (Functor m, GmState m, GmEnv m) => m (FilePath -> FilePath)
```
Can't quite decipher what this function does from the name; comment please :3

lierdakil replied:
Long story short, it creates a "reverse redirection map", i.e. redirected file name as key and filename relative to `cradleRootDir` as value. Only reason it exists is to pass it to logging/reporting code, avoiding possible multiple O(n) lookups through "forward map" (that is returned by `getMMappedFiles`)

DanielG commented on src/GHCMod.hs:
```diff
@@ -256,6 +257,14 @@ globalArgSpec =
                reqArg "OPT" $ \g o ->
                    o { ghcUserOptions = g : ghcUserOptions o }
 
+      , option "" ["map-file"] "Redirect one file to another, --map-file \"file1.hs=file2.hs\"" $
```
Could you add a more detailed user/developer facing description here as a comment, I'll expose it in the `--help` output as part of moving to optparse-applicative later.

lierdakil replied:
Short version:

CLI options:
* `--map-file "file1.hs=file2.hs"` can be used to tell ghc-mod that it should take source code for `file1.hs` from `file2.hs`. `file1.hs` can be either full path, or path relative to project root. `file2.hs` has to be either relative to current directory, or full path (preferred).
* `--map-file "file.hs"` can be used to tell ghc-mod that it should take source code for `file.hs` from stdin. File end marker is `\EOT\n`, i.e. `\x04\x0A`. `file.hs` may or may not exist, and should be either full path, or relative to project root.

Interactive commands:
* `map-file file.hs` -- tells ghc-modi to read `file.hs` source from stdin. Works the same as second form of `--map-file` CLI command.
* `unmap-file file.hs` -- unloads previously mapped file, so that it's no longer mapped. `file.hs` can be full path or relative to project root, either will work.

Exposed functions (this probably needs more thought, since exported functions are somewhat low-level):
* `getMMappedFiles` -- returns a `Map FilePath FileMapping` in `GmState` monad. This map is used for file name/source substitutions.
* `setMMappedFiles` -- allows to set said map.

Should probably export `loadMappedFile` and `unloadMappedFile` instead.

DanielG replied:
Please add that in the actual source because iirc github discards these comments (or rather makes them super hard to find) once you push over this changeset.

lierdakil replied:
I really think this should be in readme or something, not in source code comments...

DanielG replied:
It'll go into the --help output next time I rewrite the commandline parser, which is soon ;)

DanielG commented on src/GHCMod.hs:
```diff
@@ -333,7 +342,14 @@ main = handler $ do
 progMain :: (Options,[String]) -> IO ()
 progMain (globalOptions,cmdArgs) = do
-    (res,_) <- runGhcModT globalOptions $ ghcCommands cmdArgs
+    let
```
Since `setMMappedFiles` is part of the exposed API I think it would be better to always load the mappings, i.e. move this somewhere inside runGhcModT or even runGmlTWith, since the mappings aren't used unless the ghc session is active anyways.

Unless I'm misunderstanding something here. I don't really get why you'd think it's a good idea to "mutate" the read-only option environment here, that's usually a good indicator for a dirty hack :p

lierdakil replied:
This *is* a hack (kind of). Problem is, it's not possible to read from stdin while parsing options (which is needed for memory-mapped files only in context of ghc-mod standalone), so I do it here. Library calls etc don't read from stdin, and interactive code doesn't rely on parsed CLI options. So there.

DanielG replied:
How about just passing the FileMapping down to a new variant of the run* functions then?

lierdakil replied:
I see no reason to. If I just move part of this code to runGhcModT invocation, it would simplify things a fair bit.

lierdakil replied:
Here you go, no changes to globalOptions. I'd prefer to be able to ran most of this code while parsing options, but...

DanielG commented on src/GHCMod.hs:
```diff
@@ -413,9 +436,19 @@ globalCommands (cmd:_) = case cmd of
+getFileSourceFromStdin = do
+  let loop' acc = do
+        line <- getLine
+        if not (null line) && last line == '\EOT'
```
This is exactly the sort of thing that should be documented. (The EOT stuff)

DanielG commented on src/GHCMod.hs:
```diff
@@ -413,9 +436,19 @@ globalCommands (cmd:_) = case cmd of
 ghcCommands :: IOish m => [String] -> GhcModT m ()
 ghcCommands []         = fatalError "No command given (try --help)"
 ghcCommands (cmd:args) = do
+    loadMappedFiles
```
Again I think this should be part of the `run*` functions.

lierdakil replied:
Having this as part of run* functions wouldn't be wise. Essentially, this function only marshalls files specified on command line into a mutable structure. Rest of the code doesn't work with CLI options in any way (since well, neither ghc-modi, nor library has the luxury of CLI arguments)

That said, this could probably be more elegant. I'll see what I can do.

DanielG commented on Language/Haskell/GhcMod/Logger.hs:
```diff
@@ -33,6 +35,13 @@ data Log = Log [String] Builder
 
 newtype LogRef = LogRef (IORef Log)
 
+data GmPprEnv = GmPprEnv { rsDynFlags :: DynFlags
```
Just a nit-girtty convention thing: I usually use the capital letters in the type/constructor name as the prefix of the record member names, so `gpe` would be preferred.

DanielG commented:
The tests look good, you exposed the core functionality in the public api, all good. I'd like to see some more tests for the interaction with preprocessors (CPP, TH, LHS and so on) since that involves lots of temporary files being created by ghc and I can imagine some bugs lurking there.

DanielG commented:
If you want this to go into the release you'll have to fixup the remaining issues (if any) and rebase on master please :3

I'd like to do the release today if at all possible.

DanielG commented:
Okay there seem to be some nasty merge conflicts in HomeModuleGraph.hs and Target.hs, probably because of 42e72b3816ad40a2304b6bab3df4ef712909030d.

I guess this will have to go into 5.3.1.0 then, should've really warned you earlier about the impending release, I thought I did :x .

lierdakil commented:
If you can hold off till Monday at least, I'll try to fix everything we
could find and rebase during weekend. Otherwise, you'll essentially break
my plugin with this release :(
14 авг. 2015 г. 8:06 пользователь "Daniel Gröber" <notifications@github.com>
написал:

> Okay there seem to be some nasty merge conflicts in HomeModuleGraph.hs and
> Target.hs, probably because of 42e72b3
> <https://github.com/kazu-yamamoto/ghc-mod/commit/42e72b3816ad40a2304b6bab3df4ef712909030d>
> .
>
> I guess this will have to go into 5.3.1.0 then, should've really warned
> you earlier about the impending release, I thought I did :x .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kazu-yamamoto/ghc-mod/pull/484#issuecomment-130966959>
> .
>

lierdakil commented:
Apparently you went ahead and published 5.3. Oh well. Now I will be flooded with bug reports :( Anyway, I rebased on current master, and added tests for lhs, cpp and th. What I found out is that ghc can't run *any* preprocessing on files that have source in-memory. So, well, I dropped it. Now "memory mapped" files are actually written to temp. dir.

Still needs some guessing in case of nonexistent files passed to ghc-mod, but I really think this is not a huge problem.

DanielG commented:
On Sun, Aug 16, 2015 at 03:48:08PM -0700, Nikolay Yakimov wrote:

> Apparently you went ahead and published 5.3. Oh well. Now I will be
> flooded with bug reports :( 

Don't worry I'll probably do another release today because there is
still this pesky module graph caching issue that I don't want people
to run into, so I can merge it then.

> Anyway, I rebased on current master, and
> added tests for lhs, cpp and th. What I found out is that ghc can't
> run *any* preprocessing on files that have source in-memory. So, well,
> I dropped it. Now "memory mapped" files are actually written to temp.
> dir.

Ah, yeah that kinda what I thought ;)

> Still needs some guessing in case of nonexistent files passed to
> ghc-mod, but I really think this is not a huge problem.

How about canonicalizing the path as far as is possible at least?

Something like this:

```
head . rights <$> mapM (try . canonicalizePath . joinPath) $ reverse $ inits $ splitPath somePath
```

Not sure if that's much use since the last path component can still
mess it up completely if it's a symlink.

DanielG commented on test/FileMappingSpec.hs:
```diff
@@ -0,0 +1,240 @@
+              mapM_ (uncurry loadMappedFileSource) fm
+              checkSyntax ["File.lhs"]
+            res `shouldBe` "File.lhs:1:3:Warning: Top-level binding with no type signature: main :: IO ()\n"
+        -- NOTE: There is a bug in hlint that prevents it from linting lhs files.
```
Do you have an issue number for this bug?

lierdakil replied:
Only mention I found is https://code.google.com/p/ndmitchell/issues/detail?id=379. Technically, this seems to be a bug in cpphs, but I wasn't able to find mention of it (not that I looked especially hard). Suggested workaround would be passing `-XNoCPP` to `hlint`, but I think it should be possible to just run ghc `unlit` on lhs and sidestep this issue altogether.

DanielG replied:
Ah, ok. Well so long as someone knows about the issue.

lierdakil commented:
> How about canonicalizing the path as far as is possible at least?

Thought of it, but didn't get around to implementation yet. Generally speaking, if a file is suddenly changed to be a symlink during an interactive session, it will mess up things even without fancy file-mapping. So this approach might actually be good enough. During normal operation, this shouldn't be a problem.

DanielG commented:
So do you have anything else to add to the pull request? I merged it over here now https://github.com/DanielG/ghc-mod so just submit another one if so.

lierdakil commented:
Actually I do have some minor corrections, including that canonicalizing path thing we discussed, but I will only be able to do it in 12 to 14 hours from now. Should I PR directly to https://github.com/DanielG/ghc-mod?

lierdakil commented:
Nevermind, I'll PR here.

DanielG commented:
Nah I just pushed it here as well. I keep forgetting that now the tests take only like 3mins to run on travis :3
